### PR TITLE
Expose module signature when typing implementation

### PR DESCRIPTION
**Pull request by leostera**
While working on a new backend for OCaml to generate Erlang sources, I
found the need to read the .cmi file back into a `Types.signature`
value.

@Drup spotted that I was reading the file and pointed out this value was
already being read during the type checking process.

A quick check at the `Typedtree.module_coercion` showed us that it would
be difficul to extract the same information that is readily available in
the signature value.

This change will expose the signature value directly, so other backends
relying on this signature information do not need to do the extra work
of reading it again.

### CODE REVIEW

**Comment by gasche** [comment]
Naive questions:
- I can see why we would want to expose the value as a result of `typecheck_implementation`,
  but why would we want to add it as unused inputs of the other functions?
  (If this is just to preserve direct piping of one into the other, I'm not convinced.)
- If we start having n-uples with n>2, maybe we should consider using records for readability?

**Comment by leostera** [comment]
In this case I just went for the smallest change, but I'd be happy to introduce a record like:

```ocaml
type typed_impl = {
  structure: Typedtree.structure;
  coercion: Typedtree.module_coercion;
  signature: Types.signature
}
```

What do you think?

**Comment by gasche** [comment]
I think that this would be better than the current PR, yes. It doesn't really answer my question about why we want to pass the signature to the later passes (maybe because it makes sense that they could use it?), but at least it makes it completely transparent/implicit in a pleasing way.

While we are at it: I would prefer it the signature was defined and documented before the coercion, as the explanation of the coercion mentions the signature.

**Comment by leostera** [comment]
@gasche Regarding your question about passing the signature to the later passes: it doesn't make sense _right now_ for the existing later passes in the backends here. So I've deferred to pattern-matching on the record and reconstructing the tuple of values they actually use.

**Comment by leostera** [comment]
After trying to use a record defined in `Typemod` it felt a little strange that we'd be depending on _that module_ to get access to data that would otherwise come from `Typedtree`. So I moved the type there.

Name is still up for discussion, just let me know what you think :)

**Comment by gasche** [comment]
I think that the field names are natural and they work well. I initially expected the record to be defined in `compile_common`, and be used only in functions at this level (not in the "core" modules in typing/). In fact your proposal works fairly well. Just a nitpick, I would prefer `Typedtree.impl` or `Typedtree.implementation` rather than `Typedtree.t`; the latter suggests that this is "the typedtree", when it is really the typedtree of a module checked against a signature.

**Comment by leostera** [comment]
@gasche rename is in place, and CI says we're ✔️ -- let me know if you'd like anything changed.

Also @Drup let me know what you think 🙌🏼

File: typing/typedtree.mli
Comment (Drup) [inline typing/typedtree.mli:769]: This deserve a tiny bit of documentation, in particular to say that the signature is the *exported* one (i.e., might come from an `.mli`.)

**Review by Drup — APPROVED** [review]
I like the new `Typedtree.implementation` type. I see no other information to propagate right now, but it'll be useful if we ever need to, which is nice. Just need a bit of documenting and this will be good to go.

File: typing/typedtree.mli
Comment (leostera) [reply typing/typedtree.mli:769]:
```suggestion
(** A typechecked implementation including its module structure, its exported
    signature, and a coercion of the module against that signature.
  
    If an .mli file is present, the signature will come from that file and be
    the exported signature of the module.

    If there isn't one, the signature will be inferred from the module structure.
*)
```

along these lines?

Comment (trefis) [reply typing/typedtree.mli:769]: This certainly would do, though I would personally go for something more concise, like:
```ocaml
type implementation = {
  structure: structure;
  coercion: module_coercion;
  signature: Types.signature; (** exported interface of the unit, coming from the mli file if one is present, inferred from the structure otherwise *)
}
```

Comment (leostera) [reply typing/typedtree.mli:769]:
```ocaml
(** A typechecked implementation including its module structure, its exported
    signature, and a coercion of the module against that signature.
  
    The signature comes from the .mli file if one is present. Otherwise it is
    inferred from the structure. 
*)
```

How's this looking @trefis?

**Review by gasche — APPROVED** [review]
Here is a stamp of approval: this is good to go once we have converged on the documentation. Thanks for your contribution.

File: typing/typedtree.mli
Comment (trefis) [reply typing/typedtree.mli:769]: Actually, the part that I liked least (because I found it redundant), is the part that you kept.
The part that you changed, was fine (potentially better :p) in your initial version (than in mine).

But really, I don't mind :)

Comment (gasche) [reply typing/typedtree.mli:769]: The mention of `.mli/.cmi` files at this level feels a bit odd to me. It makes sense in the driver, which is concerned with the outer shell of the compiler that manipulates file, etc. But here we are in a core module, taking about types (module structure, module signatures...) that are not a-priori related to files of the filesystem.

More abstractly, I would say that this represents the result of typing a module implementation against a given module interface. We get a typedtree for the implementation (which validates that there is no error), and also a coercion from its "inner signature" to the provided interface. (In the compile_common case, the provided interface comes from the .mli if it exists.)

Which raises a new semi-naive question: in this story, there are two interfaces of interest: the "internal" interface of the module, and the "external" signature against which we are checking. If I understand correctly, the `signature` field carries around the "external" signature. Is it easy enough to recover the "internal" signature from the typed implementation, or would it make sense to also provided it here? (Without this information, the coercion does not make complete sense.)

Comment (Drup) [reply typing/typedtree.mli:769]: >  Is it easy enough to remove the "internal" signature from the typed implementation,

The inferred signature is available in the structure (`str_type` field), so it's already available.

Edit: Technically, that's not completely true, `str_type` has not been simplified to remove duplicates and so on, but it's easy enough to obtain through `Typemod.simplify`.

Comment (leostera) [reply typing/typedtree.mli:769]: @gasche I understand your point. We could maybe move this comment to `Typemod` instead where this distinction is happening, would that be more appropriate?

I get that `Typemod` is a core module, but its already relying on the existence of a `.cmi` file, so perhaps _that_ could be refactored so that we _pass in_ a signature optionally instead of letting it find it in the file system?

I'm happy to do more work to arrive at something that would be more useful / cleaner, just let me know if we want to do it in this same PR or if we want to do it separately.

Comment (gasche) [reply typing/typedtree.mli:769]: I feel that we are sort of splitting hair here, which is taking some amount of energy for small gains. If @trefis has a concrete proposal for an alternative documentation to write, we could integrate it, otherwise let's just merge the current best effort and be done with it.

**Comment by leostera** [comment]
~~Any chance this will make the cut to 4.12?~~ 

It seems I missed the deadline 😅 -- nevermind this!

**Comment by Octachron** [comment]
This is not a bugfix, thus it can wait for the next release.

**Comment by gasche** [comment]
Merged, thanks! (I forgot to squash, so the history is a bit messier, but oh well.)

**Pull request merged**
